Observe Volume Integrals

carmaza opened the pull request:
## Proposed changes

Adds event that observes volume integrals of the given tensors. The integrals will be performed component-wise for tensors of rank > 0. 

(Putting a PR in order to get feedback from people interested in having this in develop.)

### Types of changes:

- [ ] Bugfix
- [x ] New feature
- [ ] Refactor

### Component:

- [x ] Code
- [x] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [x] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [x] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

- This PR is in progress and depends on #2042 Please only check last commit. l'm currently adding some tests using the ATF.
- This action should be capable of integrating any field retrieved from the DataBox. In particular, it works with compute items, so new compute items can be added for new fields one wants to integrate.
- The action also integrates the physical volume itself: this is useful if one is interested in taking means or higher moments.
- It is not clear to me if higher moments (e.g. integrals of the square of a given field) should be handled here or elsewhere.
- This is to be compared with `ObserveErrorNorms`. Maybe in the future (or as part of this PR; I haven't thought about how to do it yet) both can be unified.

Feel free to suggest better ways to do anything here.

carmaza committed:
Add action to observe volume integrals

nilsdeppe commented on src/ParallelAlgorithms/Events/ObserveVolumeIntegrals.hpp:
```diff
@@ -0,0 +1,202 @@
+                                   const std::vector<double>& rhs) const
+        noexcept {
+      ASSERT(lhs.size() == rhs.size(),
+             "Error: vector sizes in `VectorPlus` operator do not match.");
```
No need for `Error:` could you print out the sizes?

nilsdeppe commented on src/ParallelAlgorithms/Events/ObserveVolumeIntegrals.hpp:
```diff
@@ -0,0 +1,202 @@
+    }
+  };
+
+  using VolIntDatum = Parallel::ReductionDatum<std::vector<double>, VectorPlus>;
```
I confused the `Int` here for the `int` type for a second. Maybe `VolumeIntegralDatum` or `VolIntegralDatum`?

nilsdeppe commented on src/ParallelAlgorithms/Events/ObserveVolumeIntegrals.hpp:
```diff
@@ -0,0 +1,202 @@
+  template <typename T>
+  static std::string component_suffix(const T& tensor,
+                                      size_t component_index) noexcept {
+    return tensor.rank() == 0
```
the tensor class now has a `component_suffix` member function you should use instead that does this

nilsdeppe commented on src/ParallelAlgorithms/Events/ObserveVolumeIntegrals.hpp:
```diff
@@ -0,0 +1,202 @@
+        *Parallel::get_parallel_component<observers::Observer<Metavariables>>(
+             cache)
+             .ckLocalBranch();
+    Parallel::simple_action<observers::Actions::ContributeReductionData>(
```
This is testable with the action testing framework by using the mock actions to swap in the `ContributeReductionData` and have a dummy action that just verifies everything was received that should be. My recent documentation additions to the ATF actually document and explain the mock actions a bit. We can talk about this more in person if you like.

nilsdeppe commented on src/ParallelAlgorithms/Events/ObserveVolumeIntegrals.hpp:
```diff
@@ -0,0 +1,202 @@
+        reduction_names.push_back("VolInt(" + db::tag_name<tensor_tag>() +
+                                  component_suffix(tensor, i) + ")");
+        local_vol_ints.push_back(
+            definite_integral(det_jacobian * tensor[i], mesh));
```
[not a change request] Just for the record, in the future we will want to think about and figure out how to support different types of norms, ideally controlled by the input.

nilsvu commented on src/ParallelAlgorithms/Events/ObserveVolumeIntegrals.hpp:
```diff
@@ -0,0 +1,202 @@
+      const ArrayIndex& /*array_index*/,
+      const ParallelComponent* const /*meta*/) const noexcept {
+    // Determinant of Jacobian is needed because integral is performed in
+    // logical coords. Currently not initialized in the Metavariables.
```
You could add "... as of [link to this PR]" to put the _Currently_ into context.

nilsvu commented on src/ParallelAlgorithms/Events/ObserveVolumeIntegrals.hpp:
```diff
@@ -0,0 +1,202 @@
+        get(determinant(element_map.jacobian(logical_coordinates)));
+    const double local_volume = definite_integral(det_jacobian, mesh);
+
+    std::vector<double> local_vol_ints{};
```
Also here I would prefer `local_volume_integrals` to avoid confusion with `int`

nilsvu commented on src/ParallelAlgorithms/Events/ObserveVolumeIntegrals.hpp:
```diff
@@ -0,0 +1,202 @@
+            const auto tensor_tag_v, const auto& tensor) noexcept {
+      using tensor_tag = tmpl::type_from<decltype(tensor_tag_v)>;
+      for (size_t i = 0; i < tensor.size(); ++i) {
+        reduction_names.push_back("VolInt(" + db::tag_name<tensor_tag>() +
```
Same here, name it `VolumeIntegral(...)`?

nilsvu commented on src/ParallelAlgorithms/Events/ObserveVolumeIntegrals.hpp:
```diff
@@ -0,0 +1,202 @@
+      }
+      return 0;
+    };
+    expand_pack(record_integrals(tmpl::type_<Tensors>{}, tensors)...);
```
I never understood what's the difference between `expand_pack` and macros such as `EXPAND_PACK_LEFT_TO_RIGHT`, can someone illuminate me?

nilsdeppe replied:
`expand_pack` does not guarantee that whatever is expanded into it is evaluated in any particular order, while `EXPAND_PACK_LEFT_TO_RIGHT` guarantees that things are evaluated in the order of the pack parameters.

carmaza replied:
I actually noted that in some simulations the order of the labels in the h5 file was not the order of the tensors in the typelist given in the executable. I'm changing this to `EXPAND_PACK_LEFT_TO_RIGHT` unless someone has major concerns.

nilsvu commented on src/ParallelAlgorithms/Events/ObserveVolumeIntegrals.hpp:
```diff
@@ -0,0 +1,202 @@
+  using ReductionData = tmpl::wrap<
+      tmpl::list<Parallel::ReductionDatum<double, funcl::AssertEqual<>>,
+                 Parallel::ReductionDatum<double, funcl::Plus<>>, VolIntDatum>,
+      Parallel::ReductionData>;
```
This is how you can get replace the arbitrary-sized `std::vector` in favor of a fixed-sized reduction data:

```patch
diff --git a/src/ParallelAlgorithms/Events/ObserveVolumeIntegrals.hpp b/src/ParallelAlgorithms/Events/ObserveVolumeIntegrals.hpp
index 7e7a6a2c8..d51b78510 100644
--- a/src/ParallelAlgorithms/Events/ObserveVolumeIntegrals.hpp
+++ b/src/ParallelAlgorithms/Events/ObserveVolumeIntegrals.hpp
@@ -79,27 +79,28 @@ class ObserveVolumeIntegrals<VolumeDim, ObservationValueTag,
                              tmpl::list<Tensors...>, EventRegistrars>
     : public Event<EventRegistrars> {
  private:
-  struct VectorPlus {
-    std::vector<double> operator()(const std::vector<double>& lhs,
-                                   const std::vector<double>& rhs) const
-        noexcept {
-      ASSERT(lhs.size() == rhs.size(),
-             "Error: vector sizes in `VectorPlus` operator do not match.");
-      std::vector<double> result(lhs.size());
-      for (size_t i = 0; i < lhs.size(); ++i) {
-        result[i] = lhs[i] + rhs[i];
-      }
-      return result;
-    }
-  };
-
-  using VolIntDatum = Parallel::ReductionDatum<std::vector<double>, VectorPlus>;
+  static constexpr size_t num_tensor_components =
+      tmpl::fold<tmpl::integral_list<size_t, db::item_type<Tensors>::size()...>,
+                 tmpl::size_t<0>,
+                 tmpl::plus<tmpl::_state, tmpl::_element>>::value;
 
+  using VolumeIntegralDatum = Parallel::ReductionDatum<double, funcl::Plus<>>;
   using ReductionData = tmpl::wrap<
-      tmpl::list<Parallel::ReductionDatum<double, funcl::AssertEqual<>>,
-                 Parallel::ReductionDatum<double, funcl::Plus<>>, VolIntDatum>,
+      tmpl::flatten<tmpl::list<
+          Parallel::ReductionDatum<double, funcl::AssertEqual<>>,
+          VolumeIntegralDatum,
+          tmpl::filled_list<VolumeIntegralDatum, num_tensor_components>>>,
       Parallel::ReductionData>;
 
+  template <size_t... Is>
+  static ReductionData make_reduction_data(
+      const double observation_value, const double local_volume,
+      std::array<double, num_tensor_components>&& local_volume_integrals,
+      std::index_sequence<Is...> /*meta*/) noexcept {
+    return ReductionData{observation_value, local_volume,
+                         local_volume_integrals.at(Is)...};
+  }
+
   template <typename T>
   static std::string component_suffix(const T& tensor,
                                       size_t component_index) noexcept {
@@ -157,18 +158,21 @@ class ObserveVolumeIntegrals<VolumeDim, ObservationValueTag,
         get(determinant(element_map.jacobian(logical_coordinates)));
     const double local_volume = definite_integral(det_jacobian, mesh);
 
-    std::vector<double> local_vol_ints{};
+    std::array<double, num_tensor_components> local_volume_integrals{};
     std::vector<std::string> reduction_names = {
         db::tag_name<ObservationValueTag>(), "Volume"};
-    const auto record_integrals =
-        [&local_vol_ints, &reduction_names, &det_jacobian, &mesh ](
-            const auto tensor_tag_v, const auto& tensor) noexcept {
+    size_t integral_index = 0;
+    const auto record_integrals = [&local_volume_integrals, &reduction_names,
+                                   &det_jacobian, &mesh, &integral_index](
+                                      const auto tensor_tag_v,
+                                      const auto& tensor) noexcept {
       using tensor_tag = tmpl::type_from<decltype(tensor_tag_v)>;
       for (size_t i = 0; i < tensor.size(); ++i) {
         reduction_names.push_back("VolInt(" + db::tag_name<tensor_tag>() +
                                   component_suffix(tensor, i) + ")");
-        local_vol_ints.push_back(
-            definite_integral(det_jacobian * tensor[i], mesh));
+        local_volume_integrals[integral_index] =
+            definite_integral(det_jacobian * tensor[i], mesh);
+        integral_index++;
       }
       return 0;
     };
@@ -185,8 +189,9 @@ class ObserveVolumeIntegrals<VolumeDim, ObservationValueTag,
             observation_value,
             typename Metavariables::element_observation_type{}),
         std::string{"/element_data"}, reduction_names,
-        ReductionData{static_cast<double>(observation_value), local_volume,
-                      local_vol_ints});
+        make_reduction_data(static_cast<double>(observation_value),
+                            local_volume, std::move(local_volume_integrals),
+                            std::make_index_sequence<num_tensor_components>{}));
   }
 };
 ```

nilsdeppe replied:
I think we should move this to runtime because if we want to control different norm types and variables via the input file it'll be better to have more runtime control.

nilsvu replied:
yep that sounds like the right thing to do 👍

nilsvu commented on src/ParallelAlgorithms/Events/ObserveVolumeIntegrals.hpp:
```diff
@@ -0,0 +1,202 @@
+        observers::ObservationId(
+            observation_value,
+            typename Metavariables::element_observation_type{}),
+        std::string{"/element_data"}, reduction_names,
```
The subfile name within the H5 file must be different to the one `ObserveErrorNorms` writes into, else they conflict with each other. I suggest `/volume_integrals`

nilsvu commented on src/ParallelAlgorithms/Events/ObserveVolumeIntegrals.hpp:
```diff
@@ -0,0 +1,202 @@
+                             tmpl::list<Tensors...>, EventRegistrars>
+    : public Event<EventRegistrars> {
+ private:
+  struct VectorPlus {
```
Should we move this to `Utilities/Functional.hpp`?

nilsdeppe replied:
I also thought about this and am on the fence, but I think it wouldn't hurt to have it in `Utilities/Functional.hpp` since I expect we will need to add `std::vector<T>`s more than just here.

carmaza committed:
Add Tensor component_suffix overload for storage_index

carmaza committed:
Use component_suffix from storage index overload

carmaza committed:
Move VectorPlus to Functional

carmaza commented:
Rebased on develop. I added some separate commits to address individual requests that required individual features. The commit order should now be:

- Add Tensor component_suffix overload for storage_index
- Use component_suffix from storage index overload
- Move VectorPlus to Functional
- Add action to observe volume integrals (ALREADY REVIEWED)
- fixup square(Nils) review
- fixup Add test using ATF (NEW CODE relevant to this PR)

nilsdeppe commented on tests/Unit/ParallelAlgorithms/Events/Test_ObserveVolumeIntegrals.cpp:
```diff
@@ -0,0 +1,261 @@
+  }
+};
+
+// Need forward declaration to avoid linking error.
```
This is actually constructing the static member. I think the correct comment would be `// Define static class member results`

carmaza replied:
Oh ok, thanks. Actually, I just wrote this when I was learning how to do this and I forgot to delete this comment :P I'll just delete it.

nilsdeppe replied:
Okay :) I've done the same thing a lot :P

nilsdeppe commented on tests/Unit/ParallelAlgorithms/Events/Test_ObserveVolumeIntegrals.cpp:
```diff
@@ -0,0 +1,261 @@
+
+  // ObserveVolumeIntegrals requires some domain items in the DataBox.
+  // Any domain, mesh and basis should be fine--we'll just check received data
+  const domain::creators::Brick brick({{-0.5, -0.5, -0.5}}, {{0.5, 0.5, 0.5}},
```
I *think* (haven't check carefully) this is the only place that assumes 3d. What I've done to get tests in multiple dimensions working more easily is to factor the domain creation out as:
```cpp
template <size_t Dim>
Domain<Dim> create_domain();

template <>
Domain<1> create_domain() {
  return Interval{....}.create_domain();
}

template <>
Domain<2> create_domain() {
  return Rectangle{...}.create_domain();
}

template <>
Domain<3> create_domain() {
  return Brick{....}.create_domain();
}
```
Then you should be able to factor the `SPECTRE_TEST_CASE` into a function:
```cpp
template <size_t Dim>
void test() {
  .....
}
```
and in the `SPECTRE_TEST_CASE` do:
```cpp
  test<1>();
  test<2>();
  test<3>();
```
What do you think?

carmaza replied:
I guess I can just a lambda that gives the right domain for the given dimension. The rest would be the same, as you noticed.

nilsdeppe replied:
Yea, that also works. I've found the function template specializations a bit easier to read than the `make_overloader` calls, but here it really doesn't make much difference so do whichever you prefer :)

carmaza replied:
By the way, clang-tidy is complaining about `std::move`ing `mesh`, as I am also using it to compute the expected integrals and the order of both operations is apparently not guaranteed. Should I just not move `mesh` then?

Edit: I removed most of the clang-tidy issues by not `std::move`ing the objects. I can't do that for `map` though, as the class's copy constructor was apparently `delete`d (I'll get a compiler error if I remove the `std::move`).

nilsdeppe replied:
Could you elaborate on what exactly the complaint is?

carmaza replied:
Nevermind. I removed all the clang-tidy issues by removing all the `std::move`. Should be good now.

nilsdeppe reviewed (commented):
Looks good! Couple thoughts for you to consider. I'm happy for you to squash once @nilsleiffischer has had a chance to look

nilsvu commented on src/Utilities/Functional.hpp:
```diff
@@ -308,6 +308,25 @@ struct Square : Functional<C::arity> {
+    ASSERT(lhs.size() == rhs.size(),
+           "Vector sizes in `funcl::VectorPlus` operator do not match. First "
+           "argument "
+           "size: "
```
looks like clang-format wasn't smart enough to reflow this string. Please make line 118 `"argument size:"`

nilsvu commented on tests/Unit/Utilities/Test_Functional.cpp:
```diff
@@ -468,6 +475,19 @@ SPECTRE_TEST_CASE("Unit.Utilities.Functional", "[Utilities][Unit]") {
 #endif
 }
+
+    // clang-format off
```
I didn't realize this works! good to know :) I never understood why clang-format gets confused by our error tests...

nilsvu commented on tests/Unit/ParallelAlgorithms/Events/Test_ObserveVolumeIntegrals.cpp:
```diff
@@ -0,0 +1,261 @@
+  enum class Phase { Initialization, Testing, Exit };
+
+  struct ObservationType {};
+  using element_observation_type = ObservationType;
```
I'm not sure if we really need this, and if we do, if `ObserveVolumeIntegrals` should use the same `ObservationType` as `ObserveErrorNorms`. It wouldn't hurt to define an `ObservationType` struct in `ObserveVolumeIntegrals.hpp` instead of retrieving it from the `Metavariables` I think. @nilsdeppe do you have more insight here? If there's doubt, we can just keep it as-is for now.

nilsvu commented on tests/Unit/ParallelAlgorithms/Events/Test_ObserveVolumeIntegrals.cpp:
```diff
@@ -0,0 +1,261 @@
+};
+
+template <size_t Dim>
+struct SomeSystem {
```
This is not really a `System` in the sense we use the name in other parts of the code. How about you just move the content of this struct to the anonymous namespace, where you also define `ObservationValueTag`? Then you can template the `test_observe` function on `Dim` instead.

nilsvu commented on tests/Unit/ParallelAlgorithms/Events/Test_ObserveVolumeIntegrals.cpp:
```diff
@@ -0,0 +1,261 @@
+                                        tmpl::list<>>>;
+};
+
+template <typename System>
```
This template parameter is not needed as far as I can tell

nilsvu commented on tests/Unit/ParallelAlgorithms/Events/Test_ObserveVolumeIntegrals.cpp:
```diff
@@ -0,0 +1,261 @@
+  using value_type = DataVector::value_type;
+  UniformCustomDistribution<tt::get_fundamental_type_t<value_type>> dist{-10.0,
+                                                                         10.0};
+  Variables<vars_for_test> vars(cube(brick.initial_extents()[0][0]));
```
Use `mesh.number_of_grid_points()` instead

nilsvu commented on tests/Unit/ParallelAlgorithms/Events/Test_ObserveVolumeIntegrals.cpp:
```diff
@@ -0,0 +1,261 @@
+                  "[Unit][Evolution]") {
+  using system = SomeSystem<3>;
+
+  INFO(pretty_type::get_name<system>());
```
Add scopes around these "sections" in your test so that the `INFO` only gets printed if a check within this section fails:

```cpp
{
  INFO(...);
  // check things
}
```

nilsvu reviewed (changes requested):
I'm happy with squashing existing fixups

nilsdeppe reviewed (commented):
LGTM!

@nilsleiffischer can you please take a look?

nilsvu reviewed (commented):
yep please squash 👍

carmaza commented:
Squashed and rebased on develop.

nilsdeppe reviewed (approved):
![image](https://user-images.githubusercontent.com/13531030/76226087-c21f4d80-61f3-11ea-921b-cd50d6c17e0c.png)

nilsvu commented:
Thanks for putting in the work on this PR @carmaza! This will be very helpful for @tomwlodarczyk and myself.
